modules/generic-scala3/src/main/scala/pureconfig/generic/scala3/HintsAwareProductConfigReaderDerivation.scala

```diff
@@ -0,0 +1,137 @@
+import pureconfig.generic.derivation.Utils.widen
+
+trait HintsAwareProductConfigReaderDerivation { self: HintsAwareConfigReaderDerivation =>
+  inline def deriveProductReader[A](using
+      m: Mirror.ProductOf[A],
+      ch: CoproductHint[A],
+      ph: ProductHint[A]
+  ): ConfigReader[A] =
+    inline erasedValue[A] match {
+      case _: Tuple =>
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): ConfigReader.Result[A] =
+            for {
+              listCur <- asList(cur)
+              result <- readTuple[A & Tuple, 0](listCur.list)
+            } yield result
+
+          def asList(cur: ConfigCursor) =
+            cur.asListCursor.flatMap { listCur =>
+              if (constValue[Tuple.Size[A & Tuple]] == listCur.size)
+                Right(listCur)
+              else
+                listCur.failed(
+                  WrongSizeList(constValue[Tuple.Size[A & Tuple]], listCur.size)
+                )
+            }
+        }
+
+      case _ =>
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): ConfigReader.Result[A] = {
+            val tupleSize = summonInline[ValueOf[Tuple.Size[m.MirroredElemTypes]]]
+            val defaults = ProductDerivationMacros.getDefaults[A](tupleSize.value)
+
+            for {
+              objCursor <- cur.asObjectCursor
+              labels = Utils.transformedLabels(identity)
+              actions = labels.map { label => label -> ph.from(objCursor, label) }.toMap
+              result <- readCaseClass[m.MirroredElemTypes, 0, A](objCursor, labels, actions, defaults)
+            } yield m.fromProduct(result)
+          }
+        }
+    }
+
+  private inline def readCaseClass[T <: Tuple, N <: Int, A: ProductHint: CoproductHint](
+      objCursor: ConfigObjectCursor,
+      labels: List[String],
+      actions: Map[String, ProductHint.Action],
+      defaults: Vector[Option[Any]]
+  ): Either[ConfigReaderFailures, T] =
+    inline erasedValue[T] match {
+      case _: (h *: t) =>
+        val n = constValue[N]
+        lazy val reader = summonConfigReader[h]
+        val default = defaults(n)
+        val label = labels(n)
+        val fieldHint = actions(label)
+
+        val head =
+          (fieldHint, default) match {
+            case (UseOrDefault(cursor, _), Some(defaultValue)) if cursor.isUndefined =>
+              Right(defaultValue.asInstanceOf[h])
+            case (action, _) if reader.isInstanceOf[ReadsMissingKeys] || !action.cursor.isUndefined =>
+              reader.from(action.cursor)
+            case _ =>
+              objCursor.failed(KeyNotFound.forKeys(fieldHint.field, objCursor.keys))
+          }
+
+        val tail = readCaseClass[t, N + 1, A](objCursor, labels, actions, defaults)
+
+        val resultTuple = ConfigReader.Result.zipWith(head, tail)((h, t) => widen[h *: t, T](h *: t))
+
+        val usedFields = actions.map(_._2.field).toSet
+        val hintFailures = summon[ProductHint[A]].bottom(objCursor, usedFields).toLeft(())
```

Comment:
This looks correct, but I noticed that no tests cover the behavior of `bottom`. Could you please add a test for derived readers dealing with unknown keys?

Reply:
There are tests for this, [here](https://github.com/pureconfig/pureconfig/pull/1636/files#diff-f40c59d3ad8b4e6265edbc0f6fadda20d376eeab8ab2df82be379030bd9ec767R96) and [here](https://github.com/pureconfig/pureconfig/pull/1636/files#diff-f40c59d3ad8b4e6265edbc0f6fadda20d376eeab8ab2df82be379030bd9ec767R112). 
If you'll change the this line to `val hintFailures: ConfigReader.Result[Unit] = Right(())` they will fail.

Reply:
Right, makes sense :+1: